src/algebra/hierarchy_design.lean

```diff
@@ -183,3 +186,21 @@ Another alternative to a TODO list in the doc-strings is adding github issues.
 
 -/
 library_note "the algebraic hierarchy"
+
+/--
+Some definitions that define objects of a class cannot be instances, because they have an
+explicit argument that does not occur in the conclusion. An example is `preorder.lift` that has a
+function `f : α → β` as an explicit argument to lift a preorder on `β` to a preorder on `α`.
+
+If these definitions are used to define other instances *and* it is important that these other
+instances will be unfolded by type-class inference, then these definitions should be marked
+`@[reducible]`.
```

Comment:
Doesn't typeclass inference already do this unfolding though? Is this just about providing a fast path to unfold the obvious noise first? Is the point in this unfolding to stop typeclass inference attempting to unify `f` in this example before further unfolding?

Reply:
No, type-class inference doesn't unfold semireducible (default) definitions, only instances and reducible definitions.